common/unified/solver/chebyshev_kernels.cpp

```diff
@@ -0,0 +1,114 @@
+// SPDX-FileCopyrightText: 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "core/solver/chebyshev_kernels.hpp"
+
+#include <type_traits>
+
+#include <ginkgo/core/matrix/dense.hpp>
+
+#include "common/unified/base/kernel_launch.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace GKO_DEVICE_NAMESPACE {
+namespace chebyshev {
+
+
+template <typename T>
+using coeff_type = gko::kernels::chebyshev::coeff_type<T>;
+
+
+#if GINKGO_DPCPP_SINGLE_MODE
+
+
+// we only change type in device code to keep the interface is the same as the
+// other backend.
+template <typename coeff_type>
+using if_single_only_type =
+    std::conditional_t<std::is_same_v<coeff_type, double>, float,
+                       std::complex<float>>;
+
+
+#else
+
+
+template <typename T>
+struct type_identity {
+    using type = T;
+};
+
+
+template <typename coeff_type>
+using if_single_only_type = typename type_identity<coeff_type>::type;
+
+
+#endif
+
+
+template <typename ValueType>
+void init_update(std::shared_ptr<const DefaultExecutor> exec,
+                 const coeff_type<ValueType> alpha,
+                 const matrix::Dense<ValueType>* inner_sol,
+                 matrix::Dense<ValueType>* update_sol,
+                 matrix::Dense<ValueType>* output)
+{
+    using actual_coeff_type = if_single_only_type<coeff_type<ValueType>>;
+    using type = device_type<actual_coeff_type>;
+
+    auto alpha_val = static_cast<actual_coeff_type>(alpha);
+
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto row, auto col, auto alpha, auto inner_sol,
+                      auto update_sol, auto output) {
+            const auto inner_val = static_cast<type>(inner_sol(row, col));
+            update_sol(row, col) =
+                static_cast<device_type<ValueType>>(inner_val);
+            output(row, col) = static_cast<device_type<ValueType>>(
+                static_cast<type>(output(row, col)) +
+                static_cast<type>(alpha) * inner_val);
```

Comment:
I think this has a couple unnecessary casts. The `run_kernel` should apply `as_device_type` to all arguments, so casting `static_cast<type>(alpha)` should be unnecessary.
```suggestion
            update_sol(row, col) = inner_sol(row, col);
            output(row, col) = static_cast<device_type<ValueType>>(
                static_cast<type>(output(row, col)) +
                alpha * static_cast<type>(inner_sol(row, col)));
```

Reply:
I don't understand your point, there is no mixed precision here, there is only one value type.

Reply:
you are right. I thought I do not use the truth of double/complex<double> is also the arithmetic type.
type is more on the arithmetic type so the other needs to casting to the highest precision.
delete the casting for the type but I will change the type to arithmetic_type not device_coeff_type for the purpose